### PR TITLE
Build versioning

### PR DESCRIPTION
**Pull request by paha**
RE: #45
- Added version to cli and server packages
- Version is injected by make based either on a Env VAR or
autoincremented patch level
- If version is not injected during build we panic
- For testing a simple string is added
- Added unit tests
- Very minor clean up and style improvement

TODO:
- Update documentation
- Additional testing

### CODE REVIEW

**Comment by SleepyBrett** [comment]
Hey pavel thanks for the work, I'll try and take a look at this sometime today. I certainly wanted to improve the versioning handling!

File: Makefile [hunk 1]
```diff
@@ -8,7 +8,10 @@ GITHUB_REPO_HOST_AND_PATH := github.com/$(GITHUB_REPO_OWNER)/$(GITHUB_REPO_NAME)
 IMAGE_NAME := quay.io/nordstrom/kubelogin
 BUILD := build
 CURRENT_BRANCH := $(shell git rev-parse --abbrev-ref HEAD)
-CURRENT_TAG := v0.0.52
+# NOTE: only patch level version is auto incremented to set a new version
+# 		to specify a new version: $ make version CURRENT_TAG="v0.2.0"
+LAST_TAG := $(shell git tag --list | tail -1)
+CURRENT_TAG ?= $(shell echo $(LAST_TAG) | awk -F. '{print $$1"."$$2"."$$NF+1}')
```
Comment (mariusgrigoriu) [inline Makefile]: I think it isn't necessary or desirable to auto-increment patch level here. The release workflow on GitHub is to go to Releases > Draft new release. From there you enter in a tag such as v0.0.52 and GitHub will apply that tag. So the version is all manually driven anyway.

File: Makefile [hunk 2]
```diff
@@ -160,3 +165,8 @@ build $(BUILD)/server $(BUILD)/server/linux $(BUILD)/cli $(BUILD)/cli/mac $(BUIL
 
 clean:
 	rm -rf build
+
```
Comment (mariusgrigoriu) [inline Makefile]: Is this left over from debugging? Seems to serve no purpose.

File: cmd/cli/app.go [hunk 3]
```diff
@@ -0,0 +1,118 @@
+package main
```
Comment (mariusgrigoriu) [inline cmd/cli/app.go]: Thank you for breaking our code out from a single file.

File: cmd/cli/main.go [hunk 4]
```diff
@@ -179,164 +118,76 @@ func setFlags(command *flag.FlagSet, loginCmd bool) {
...
-			if loginCommmand.Parsed() {
-				if kubeloginServerBaseURL == "" {
-					log.Fatal("--server-url must be set!")
-				}
-				app.kubectlUser = userFlag
-				app.kubeloginServer = kubeloginServerBaseURL
-				generateURLAndListenForServerResponse(app)
-			}
-		}
+		login(app)
```
Comment (mariusgrigoriu) [inline cmd/cli/main.go]: Thank you for extracting into a function. Makes it much more readable.

File: cmd/cli/main.go [hunk 5]
```diff
@@ -179,164 +118,76 @@ func setFlags(command *flag.FlagSet, loginCmd bool) {
...
-	configCommand := flag.NewFlagSet("config", flag.ExitOnError)
-	setFlags(configCommand, false)
+
+	fortest() // Check if test are run
+
 	user, err := user.Current()
 	if err != nil {
 		log.Fatalf("Could not determine current user of this system. Err: %v", err)
 	}
+
```
Comment (mariusgrigoriu) [inline cmd/cli/main.go]: Seems out of place given the switch statement below. Use the same approach as login or config?

File: cmd/cli/main_test.go [hunk 6]
```diff
@@ -11,6 +11,17 @@ import (
 	. "github.com/smartystreets/goconvey/convey"
 )
 
+func init() {
+	testTest = true
+}
+
+func TestVersion(t *testing.T) {
```
Comment (mariusgrigoriu) [inline cmd/cli/main_test.go]: The version number is a function of the makefile and compiler flags. Please help me see the need for this test and the `forest()` function.

**Pull request closed without merging**